### PR TITLE
Refactor some components / screens to use `useOnyx`

### PR DESCRIPTION
**Pull request by fabioh8010**
### Details

This PR refactors some components / screens in E/App to use `useOnyx` hook and assert it's behaving correctly.

Components / screens refactored:

- src/pages/NewChatPage.tsx - example with `initWithStoredValues`
- src/pages/ProfilePage.tsx - example with `selector`, dynamic key
- src/pages/settings/Preferences/PreferencesPage.tsx
- src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.tsx - example with `selector`, `initialValue`, `isLoadingOnyxValue` function usage

### Fixed Issues

$ https://github.com/Expensify/App/issues/34339
PROPOSAL:

#### NewChatPage

1. Open "Start chat" page.
2. Refresh the page and ensure transitions and loading are working as expected (web / mobile web).
3. Select a contact to start a chat.
4. Open "Start chat" page again.
5. Type something in the search and wait for the loading to finish.
6. Type something again and refresh the page while the loading is ongoing, ensure the loading disappears after the refresh (web / mobile web).

#### ProfilePage

1. Click on someone's avatar to open the Profile page.
2. Refresh the page and ensure transitions and loading are working as expected (web / mobile web).
3. Change notification option, open "Private notes" page, click "Message..." button, ensure the UI is updated and everything is working as expected.

#### PreferencesPage

1. Navigate to Preferences page.
2. Toggle first Notifications option, refresh the page and ensure transitions and loading are working as expected (web / mobile web).
3. Toggle the other options, change all settings in the page, ensure the UI is updated and everything is working as expected.

#### ContactMethodDetailsPage

1. Navigate to "Contact methods" page.
2. Ensure you have at least 2 contact methods set up.
3. Click the non-default contact method to open the details page.
4. Refresh the page and ensure transitions and loading are working as expected (web / mobile web).
5. Set the contact method as default, ensure the process makes any effect (even if it fails).
6. Ensure the "Contact methods" page's UI is updated.

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

Same as *Tests*

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<h4>NewChatPage</h4>

https://github.com/Expensify/App/assets/20051562/1ef67f7c-8ebc-453b-857a-454f7f3c7d80

<h4>ProfilePage</h4>

https://github.com/Expensify/App/assets/20051562/bdb053f0-c599-45fb-a46e-dc7227a0cbeb

<h4>PreferencesPage</h4>

https://github.com/Expensify/App/assets/20051562/f13c4e54-e945-42a3-96bc-84c4d299b8f4

<h4>ContactMethodDetailsPage</h4>

https://github.com/Expensify/App/assets/20051562/87871c90-b8d5-47d9-9f02-0f224c22c1ed

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<h4>NewChatPage</h4>

https://github.com/Expensify/App/assets/20051562/9d160af7-bfa7-4634-b628-7cd89a69378b

<h4>ProfilePage</h4>

https://github.com/Expensify/App/assets/20051562/2d574f72-2bd4-4a07-a23f-4bc6f04fb265

<h4>PreferencesPage</h4>

https://github.com/Expensify/App/assets/20051562/ef6a2f99-8797-4362-bd4c-a490ef1a5eee

<h4>ContactMethodDetailsPage</h4>

https://github.com/Expensify/App/assets/20051562/8f06fa22-bab6-4572-a95c-81dcf60ecaef

</details>

<details>
<summary>iOS: Native</summary>

<h4>NewChatPage</h4>

https://github.com/Expensify/App/assets/20051562/6047bf1f-2aab-4eb6-b301-7ef88617e741

<h4>ProfilePage</h4>

https://github.com/Expensify/App/assets/20051562/096d5e6c-a30d-4e4a-8fc0-618ae41460a9

<h4>PreferencesPage</h4>

https://github.com/Expensify/App/assets/20051562/49d34273-f640-42de-bd9e-9dd2254ac177

<h4>ContactMethodDetailsPage</h4>

https://github.com/Expensify/App/assets/20051562/10d92997-7d05-46cf-af62-670108f6e8e9

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<h4>NewChatPage</h4>

https://github.com/Expensify/App/assets/20051562/9eb44874-9fca-43a4-96a6-36835a17895b

<h4>ProfilePage</h4>

https://github.com/Expensify/App/assets/20051562/096ae707-9ae5-40dc-815d-56ad43cf7e9b

<h4>PreferencesPage</h4>

https://github.com/Expensify/App/assets/20051562/663af367-bc38-425a-b24c-96e2e5a3387c

<h4>ContactMethodDetailsPage</h4>

https://github.com/Expensify/App/assets/20051562/d9fc72cf-921e-4ca2-a0d2-c69c44597975

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<h4>NewChatPage</h4>

https://github.com/Expensify/App/assets/20051562/deb6beb3-9933-4669-bc4e-8117de8b4752

https://github.com/Expensify/App/assets/20051562/6c7e2d99-74b6-4216-8b67-c229da23a059

<h4>ProfilePage</h4>

https://github.com/Expensify/App/assets/20051562/c18757b1-51e9-40df-8296-925b95160193

https://github.com/Expensify/App/assets/20051562/0def6993-5db5-4a20-9251-07fb756a2059

<h4>PreferencesPage</h4>

https://github.com/Expensify/App/assets/20051562/945b5e81-f685-4f98-b97c-62fecfc68562

https://github.com/Expensify/App/assets/20051562/feaf24e3-8af7-40fc-902c-1ab384df64bb

<h4>ContactMethodDetailsPage</h4>

https://github.com/Expensify/App/assets/20051562/b802ac4a-25e9-4728-bb0a-db779f63168a

https://github.com/Expensify/App/assets/20051562/2aa1f617-9937-4ced-9a94-d6c90244fc27

</details>

<details>
<summary>MacOS: Desktop</summary>

<h4>NewChatPage</h4>

https://github.com/Expensify/App/assets/20051562/d0726cea-b4d5-4701-b44c-4749d4ef0846

<h4>ProfilePage</h4>

https://github.com/Expensify/App/assets/20051562/28577aaa-03b8-4c07-98ef-c0c878c368d7

<h4>PreferencesPage</h4>

https://github.com/Expensify/App/assets/20051562/36f1a7bc-8eda-4337-bdb9-f310576d0e74

<h4>ContactMethodDetailsPage</h4>

https://github.com/Expensify/App/assets/20051562/4d89263d-4b50-41a7-a102-4ab9c3cdb0e4

</details>

### CODE REVIEW

**Comment by fabioh8010** [comment]
**Update:** I'm currently having an issue with `ContactMethodDetailsPage`. When refactoring it to `useOnyx`, it displays a strange navigation behavior when reloading the page, the URL changes briefly and the screen is opened late.

## withOnyx

https://github.com/Expensify/App/assets/20051562/7b82d053-e2bd-4830-8c91-ae20df087a62

## useOnyx

https://github.com/Expensify/App/assets/20051562/8cc83233-0e6b-4d99-b845-eab0e343eb3b

---

Among the screens I refactored, this is the only one that displays this behavior, not sure why yet.

File: src/pages/ProfilePage.tsx
Comment (blazejkustra) [inline src/pages/ProfilePage.tsx:80]: How about a little change here? Instead of return type I would add as const to return statements. OR extract this type to a separate type?

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.tsx
Comment (blazejkustra) [inline src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.tsx:45]: How about a rule to always define selectors as separate function? And reuse exactly like with `chatReportSelector` function

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.tsx
Comment (blazejkustra) [inline src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.tsx:169]: Sleek! 👌

File: src/types/utils/isLoadingOnyxValue.ts
Comment (blazejkustra) [inline src/types/utils/isLoadingOnyxValue.ts:3]: Let's improve this type here, I think we should export `ResultMetadata` from onyx

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.tsx
Comment (fabioh8010) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.tsx]: I will extract to a separate function but we shouldn't take this as a rule, because I would want to use some component's data inside that selector, e.g.:

```tsx
    const [isLoadingReportData, isLoadingReportDataResult] = useOnyx(ONYXKEYS.IS_LOADING_REPORT_DATA, {initialValue: true});
    const [policies, policiesResult] = useOnyx(ONYXKEYS.COLLECTION.POLICY, {
        selector: (data) => ({
            id: isLoadingReportData ? data?.id ?? '' : 'something', // <- using external data inside selector
            ownerAccountID: data?.ownerAccountID,
            owner: data?.owner ?? '',
        }),
    });
```

File: src/types/utils/isLoadingOnyxValue.ts
Comment (fabioh8010) [reply src/types/utils/isLoadingOnyxValue.ts:3]: Will raise a PR in Onyx!

**Comment by fabioh8010** [comment]
**Update:** Addressed some review comments and figured it out the reason for [this issue](https://github.com/Expensify/App/pull/38924#issuecomment-2017782553), working on it. cc @blazejkustra @roryabraham

File: src/types/utils/isLoadingOnyxValue.ts
Comment (blazejkustra) [reply src/types/utils/isLoadingOnyxValue.ts:3]: Can you link the PR for this? @fabioh8010

Comment (blazejkustra) [inline src/types/utils/isLoadingOnyxValue.ts:3]: Also I'm wondering can we implement some kind of type narrowing to ensure that onyx values are defined? This way we could avoid optional chaining and nullish coalescing inside of components. Just an idea, we don't have to do it in this PR

**Review by blazejkustra — APPROVED** [review]
Code looks good! Let's wait for the onyx bump before merging 🚀

File: src/types/utils/isLoadingOnyxValue.ts
Comment (fabioh8010) [reply src/types/utils/isLoadingOnyxValue.ts:3]: Sorry, forgot about it! Here is it: https://github.com/Expensify/react-native-onyx/pull/527

But I think we shouldn't hold this PR because of this, otherwise we will require another bump of multiple versions that would take more time.

Comment (fabioh8010) [reply src/types/utils/isLoadingOnyxValue.ts:3]: But the thing is that the Onyx value can still be `null` even if it's loaded. This helper function is useful for check if all values were loaded into cache and ready to be used, but it doesn't mean they are not `null`.

I previously thought about using type narrowing to consider the `status` off all the values to `loaded` if the condition is true, but not sure if it's really useful neither feasible.

**Comment by fabioh8010** [comment]
@blazejkustra @roryabraham 

Removing it from HOLD because #38114 was merged 🎉 Changes were already applied on this PR.

**Review by blazejkustra — APPROVED** [review]

**Comment by fabioh8010** [comment]
@ishpaul777 I believe is related to [this issue](https://github.com/Expensify/App/issues/39642), backend issues with the search results.

**Review by ishpaul777 — APPROVED** [review]

**Comment by fabioh8010** [comment]
@roryabraham bump

**Review by roryabraham — APPROVED** [review]
it's beautiful 😭

**Comment by roryabraham** [comment]
long live `useOnyx`

**Pull request merged**
